packages/zoe/src/zoeService/installationStorage.js

```diff
@@ -80,54 +72,79 @@ export const makeInstallationStorage = (
+    installBundleID: M.call(M.string()).returns(M.promise()),
+    unwrapInstallation: M.callWhen(M.eref(InstallationShape)).returns(M.any()),
+    getBundleIDFromInstallation: M.call(InstallationShape).returns(M.promise()),
+  });
 
-  /** @type {UnwrapInstallation} */
-  const unwrapInstallation = installationP => {
-    return E.when(installationP, installation => {
-      if (installationsBundleCap.has(installation)) {
-        const { bundleCap, bundleID } =
-          installationsBundleCap.get(installation);
-        return { bundleCap, bundleID, installation };
-      } else if (installationsBundle.has(installation)) {
-        const bundle = installationsBundle.get(installation);
-        return { bundle, installation };
-      } else {
-        assert.fail(X`${installation} was not a valid installation`);
-      }
-    });
-  };
+  const installationStorage = vivifyFarInstance(
+    zoeBaggage,
+    'InstallationStorage',
+    InstallationStorage,
+    {
+      async installBundle(allegedBundle) {
+        // Bundle is a very open-ended type and we must decide here
+        // whether to treat it as either a HashBundle or SourceBundle.
+        // So, we cast it down and inspect it.
+        const bundle = /** @type {unknown} */ (harden(allegedBundle));
+        assert.typeof(bundle, 'object', 'a bundle must be provided');
+        assert(bundle !== null, 'a bundle must be provided');
+        const { moduleFormat } = bundle;
+        if (moduleFormat === 'endoZipBase64Sha512') {
+          const { endoZipBase64Sha512 } = bundle;
+          assert.typeof(
+            endoZipBase64Sha512,
+            'string',
+            `bundle endoZipBase64Sha512 must be a string, got ${endoZipBase64Sha512}`,
+          );
+          return this.installBundleID(`b1-${endoZipBase64Sha512}`);
+        }
+        return installSourceBundle(bundle);
+      },
+      async installBundleID(bundleID) {
+        assert.typeof(bundleID, 'string', `a bundle ID must be provided`);
+        // this waits until someone tells the host application to store the
+        // bundle into the kernel, with controller.validateAndInstallBundle()
+        const bundleCap = await getBundleCapForID(bundleID);
+        // AWAIT
 
-  const getBundleIDFromInstallation = async allegedInstallationP => {
-    const { bundleID } = await unwrapInstallation(allegedInstallationP);
-    // AWAIT
-    assert(bundleID, 'installation does not have a bundle ID');
-    return bundleID;
-  };
+        /** @type {Installation} */
+        // @ts-expect-error cast
+        const installation = makeBundleIDInstallation();
+        installationsBundleCap.init(
+          installation,
+          harden({ bundleCap, bundleID }),
+        );
+        return installation;
+      },
+      unwrapInstallation(installationP) {
+        return E.when(installationP, installation => {
+          if (installationsBundleCap.has(installation)) {
```

Comment:
See comment on `callWhen` above

```suggestion
      unwrapInstallation(installation) {
        if (installationsBundleCap.has(installation)) {
```